src/pymor/algorithms/ml/vkoga.py

```diff
@@ -0,0 +1,426 @@
+        the current subspace can represent :math:`k(\,\cdot\,,x)`, i.e. the projection
+        error in the reproducing kernel Hilbert space):
+        The power function in matrix form is defined as
+
+        .. math::
+            \begin{align*}
+                P_n^2(x) = k(x,x) - k(x,X_n)k(X_n,X_n)^{-1}k(X_n,x).
+            \end{align*}
+
+        Since we would like to use the power function as selection criterion for centers
+        within the greedy iteration, we consider a scalar version by taking the trace:
+
+        .. math::
+            \begin{align*}
+                p_n^2(x) = \operatorname{trace}(P_n^2(x)).
+            \end{align*}
+
+        We are going to track this quantity evaluated at all training points during the
+        iterations of the greedy algorithm. In order to do so, we also maintain an array
+        storing the Newton basis evaluations of the current basis at the training points:
+
+        .. math::
+            \begin{align*}
+                V_{n,i} = L_n^{-1}k(X_n,y_i)
+            \end{align*}
+
+        for all training points :math:`y_i`.
+
+        Given the first center :math:`x_1`, we initialize the (squared) power function
+        values as
+
+        .. math::
+            \begin{align*}
+                p_n^2(y_i) = \operatorname{trace}(k(y_i,y_i)-V_{1,i}^\top V_{1,i}),
+            \end{align*}
+
+        where :math:`V_{1,i} = L_{22}^{-1}k(x_1,y_i)`.
+
+        The incremental updates of :math:`V_{n,i}` to :math:`V_{n+1,i}` and the power
+        function values is then performed in the following way:
+        Define :math:`\Delta_i=k(x_{n+1},y_i)-W^\top V_i` :math:`\Xi=L_{22}^{-1}\Delta_i`.
+        Then, we have
+
+        .. math::
+            \begin{align*}
+                p_{n+1}^2(y_i) = p_n^2(y_i) - \lVert \Xi_i\rVert_F^2
+            \end{align*}
+
+        and
+
+        .. math::
+            \begin{align*}
+                V_{n+1,i} = \begin{bmatrix}
+                    V_{n,i} \\
+                    \Xi_i
+                \end{bmatrix}.
+            \end{align*}
+
+        Parameters
+        ----------
+        mu
+            Parameter (center) to add.
+        """
+        mu = np.asarray(mu)
+
+        # --- Find corresponding training index (exact or nearest match) ---
+        # this is mainly required since we do not enforce that `mu` is in the training set;
+        # when introducing a suitable `Dataset` class, this should be handled by the dataset
+        matches = np.where(np.all(np.isclose(self.X_train, mu), axis=1))[0]
+        if matches.size > 0:
+            idx_in_X = int(matches[0])
+        else:
+            idx_in_X = int(np.argmin(np.linalg.norm(self.X_train - mu, axis=1)))
+
+        f_new = self.F_train[idx_in_X].reshape(self.m,)
+
+        # --- Case 1: first center ---
+        if self._L is None or self._centers is None:
+            K_nn = self.kernel(mu, mu) + self.reg * np.eye(self.m)
+            L22 = np.linalg.cholesky(K_nn)
```

Comment:
In the case `L_22` is (if scalar) the square root of the 1x1 kernel matrix, so i find the name with 22 in the index a bit confusing.